### PR TITLE
feat: AI Proxy Wasm Plugin Integration with GitHub Models #1304

### LINKED ISSUE #1304: AI Proxy Wasm Plugin Integration with GitHub Models
https://docs.github.com/en/github-models/prototyping-with-ai-models
https://github.com/marketplace/models/azure-openai/gpt-4o

```bash
curl -X POST "https://models.inference.ai.azure.com/chat/completions" \
    -H "Content-Type: application/json" \
    -H "Authorization: Bearer $GITHUB_TOKEN" \
    -d '{
        "messages": [
            {
                "role": "system",
                "content": "You are a helpful assistant."
            },
            {
                "role": "user",
                "content": "What is the capital of France?"
            }
        ],
        "model": "gpt-4o"
    }'
```

ref: https://github.com/alibaba/higress/issues/940

### PR DESCRIPTION
**Pull request by hackerYHT**
### Ⅰ. Describe what this PR did

AI Proxy Wasm Plugin Integration with GitHub Models #1304 

API documentation: https://github.com/marketplace/models/azure-openai/gpt-4o

### Ⅱ. Does this pull request fix one issue?

fixes #1304 

### Ⅳ. Describe how to verify it

openai协议，非流式

`curl -X POST "http://127.0.0.1:10000/v1/chat/completions" \
    -H "Content-Type: application/json" \
    -d '{
        "messages": [
            {
                "role": "system",
                "content": "You are a helpful assistant."
            },
            {
                "role": "user",
                "content": "What is the capital of France?"
            }
        ],
        "temperature": 1.0,
        "top_p": 1.0,
        "max_tokens": 1000,
        "model": "gpt-4o"
    }'`

回复

`{
  "choices": [
    {
      "finish_reason": "stop",
      "index": 0,
      "logprobs": null,
      "message": {
        "content": "The capital of France is Paris.",
        "role": "assistant"
      }
    }
  ],
  "created": 1728131051,
  "id": "chatcmpl-AEy7PU2JImdsD1W6Jw8GigZSEnM2u",
  "model": "gpt-4o-2024-08-06",
  "object": "chat.completion",
  "system_fingerprint": "fp_67802d9a6d",
  "usage": {
    "completion_tokens": 7,
    "prompt_tokens": 24,
    "total_tokens": 31
  }
}`

openai协议，流式

`curl -X POST "http://127.0.0.1:10000/v1/chat/completions" \
    -H "Content-Type: application/json" \
    -d '{
        "messages": [
            {
                "role": "system",
                "content": "You are a helpful assistant."
            },
            {
                "role": "user",
                "content": "What is the capital of France?"
            }
        ],
        "stream": true,
        "temperature": 1.0,
        "top_p": 1.0,
        "max_tokens": 1000,
        "model": "gpt-4o"
    }'`

回复

`data: {"choices":[{"delta":{"role":"assistant"},"finish_reason":null,"index":0,"logprobs":null}],"created":1728132827,"id":"chatcmpl-AEya3ZysGvKnIsYoPwE1lJNmDXQqZ","model":"gpt-4o-2024-08-06","object":"chat.completion.chunk","system_fingerprint":"fp_67802d9a6d"}

data: {"choices":[{"delta":{"content":"The"},"finish_reason":null,"index":0,"logprobs":null}],"created":1728132827,"id":"chatcmpl-AEya3ZysGvKnIsYoPwE1lJNmDXQqZ","model":"gpt-4o-2024-08-06","object":"chat.completion.chunk","system_fingerprint":"fp_67802d9a6d"}

data: {"choices":[{"delta":{"content":" capital"},"finish_reason":null,"index":0,"logprobs":null}],"created":1728132827,"id":"chatcmpl-AEya3ZysGvKnIsYoPwE1lJNmDXQqZ","model":"gpt-4o-2024-08-06","object":"chat.completion.chunk","system_fingerprint":"fp_67802d9a6d"}

data: {"choices":[{"delta":{"content":" of"},"finish_reason":null,"index":0,"logprobs":null}],"created":1728132827,"id":"chatcmpl-AEya3ZysGvKnIsYoPwE1lJNmDXQqZ","model":"gpt-4o-2024-08-06","object":"chat.completion.chunk","system_fingerprint":"fp_67802d9a6d"}

data: {"choices":[{"delta":{"content":" France"},"finish_reason":null,"index":0,"logprobs":null}],"created":1728132827,"id":"chatcmpl-AEya3ZysGvKnIsYoPwE1lJNmDXQqZ","model":"gpt-4o-2024-08-06","object":"chat.completion.chunk","system_fingerprint":"fp_67802d9a6d"}

data: {"choices":[{"delta":{"content":" is"},"finish_reason":null,"index":0,"logprobs":null}],"created":1728132827,"id":"chatcmpl-AEya3ZysGvKnIsYoPwE1lJNmDXQqZ","model":"gpt-4o-2024-08-06","object":"chat.completion.chunk","system_fingerprint":"fp_67802d9a6d"}

data: {"choices":[{"delta":{"content":" Paris"},"finish_reason":null,"index":0,"logprobs":null}],"created":1728132827,"id":"chatcmpl-AEya3ZysGvKnIsYoPwE1lJNmDXQqZ","model":"gpt-4o-2024-08-06","object":"chat.completion.chunk","system_fingerprint":"fp_67802d9a6d"}

data: {"choices":[{"delta":{"content":"."},"finish_reason":null,"index":0,"logprobs":null}],"created":1728132827,"id":"chatcmpl-AEya3ZysGvKnIsYoPwE1lJNmDXQqZ","model":"gpt-4o-2024-08-06","object":"chat.completion.chunk","system_fingerprint":"fp_67802d9a6d"}

data: {"choices":[{"delta":{},"finish_reason":"stop","index":0,"logprobs":null}],"created":1728132827,"id":"chatcmpl-AEya3ZysGvKnIsYoPwE1lJNmDXQqZ","model":"gpt-4o-2024-08-06","object":"chat.completion.chunk","system_fingerprint":"fp_67802d9a6d"}

data: [DONE]`

openai协议，请求embedding模型

curl -X POST "http://127.0.0.1:10000/v1/embeddings" \
    -H "Content-Type: application/json" \
    -d '{
        "input": ["first phrase", "second phrase", "third phrase"],
        "model": "text-embedding-3-large"
    }'

回复

`{
  "object": "list",
  "data": [
    {
      "object": "embedding",
      "index": 0,
      "embedding": [
        -0.0012583479,
        0.0020349282,
        ...
        0.012051377,
        -0.0053306012,
        0.0060688322
      ]
    }
  ],
  "model": "text-embedding-3-large",
  "usage": {
    "prompt_tokens": 6,
    "total_tokens": 6
  }
}`

### CODE REVIEW

File: plugins/wasm-go/extensions/ai-proxy/README.md [hunk 1]
```diff
@@ -143,6 +143,10 @@ Groq 所对应的 `type` 为 `groq`。它并无特有的配置字段。
 
 360智脑所对应的 `type` 为 `ai360`。它并无特有的配置字段。
 
+#### github模型
+
+github所对应的 `type` 为 `ai360`。它并无特有的配置字段。
```
Comment (CH3CHO) [inline plugins/wasm-go/extensions/ai-proxy/README.md:148]: GitHub 大小写要按照官方的来。另外 type 取值不对吧？

File: plugins/wasm-go/extensions/ai-proxy/README.md [hunk 2]
```diff
@@ -1018,6 +1022,105 @@ provider:
 }
 ```
 
+### 使用 OpenAI 协议代理github 模型服务
```
Comment (CH3CHO) [inline plugins/wasm-go/extensions/ai-proxy/README.md:1025]: GitHub 大小写问题同上。另外这里的空格规则是什么，英文单词前面不空格后面空格？

File: plugins/wasm-go/extensions/ai-proxy/README.md [hunk 3]
```diff
@@ -1018,6 +1022,105 @@ provider:
...
+
+**配置信息**
+
+```yaml
+provider:
+  type: github
+  apiTokens:
+    - "YOUR_MINIMAX_API_TOKEN"
```
Comment (CH3CHO) [inline plugins/wasm-go/extensions/ai-proxy/README.md:1033]: 这个占位值不对吧？

File: plugins/wasm-go/extensions/ai-proxy/provider/github.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	"errors"
+	"fmt"
+	"github.com/higress-group/proxy-wasm-go-sdk/proxywasm"
+	"github.com/higress-group/proxy-wasm-go-sdk/proxywasm/types"
+
+	"github.com/alibaba/higress/plugins/wasm-go/extensions/ai-proxy/util"
+	"github.com/alibaba/higress/plugins/wasm-go/pkg/wrapper"
+)
+
+// githubProvider is the provider for 360 OpenAI service.
```
Comment (CH3CHO) [inline plugins/wasm-go/extensions/ai-proxy/provider/github.go:14]: 蛤？

**Comment by CH3CHO** [comment]
这里整合的其实是 Azure OpenAI 模型，叫 GitHub 模型是不是言过其实了？用 OpenAI 加自定义 URL 能完成对接吗？

**Review by CH3CHO — CHANGES_REQUESTED** [review]

**Comment by hackerYHT** [comment]
> 这里整合的其实是 Azure OpenAI 模型，叫 GitHub 模型是不是言过其实了？用 OpenAI 加自定义 URL 能完成对接吗？

github的model market的模型都是请求https://models.inference.ai.azure.com 这个域名，只是在请求体当中指定“ model”。这里是可以请求github model market提供的所有模型，不单单是指 Azure OpenAI 模型。

**Comment by CH3CHO** [comment]
> > 这里整合的其实是 Azure OpenAI 模型，叫 GitHub 模型是不是言过其实了？用 OpenAI 加自定义 URL 能完成对接吗？
> 
> github的model market的模型都是请求https://models.inference.ai.azure.com 这个域名，只是在请求体当中指定“ model”。这里是可以请求github model market提供的所有模型，不单单是指 Azure OpenAI 模型。

OK

File: plugins/wasm-go/extensions/ai-proxy/provider/github.go (hunk 4 above)
Comment (hackerYHT) [reply plugins/wasm-go/extensions/ai-proxy/provider/github.go:14]: 不好意思，上述问题没有注意到，我这边修改一下

**Review by CH3CHO — APPROVED** [review]
LGTM

**Pull request merged**